### ISSUE TITLE
Undocumented change of behavior for Embeddings in PyTorch 1.8

### ISSUE THREAD

**Issue by sgugger** [pytorch/pytorch#53368]
## 📚 Documentation

I'm filing this as a documentation problem but it could also be a bug if the documentation is what you want to be right.

Starting PyTorch 1.8, the `Embedding` layer will always output a vector of zeros for the `padding_idx` used to create it. Here is a short snippet to observe the behavior:

```python
import torch
import torch.nn as nn

embeds = nn.Embedding(5, 12, padding_idx=0) # Weights at 0 are initialized with zeros
embeds.weight.data[0] = torch.ones(12) # Put another value
embeds(torch.tensor([0]))
```

In Pytorch <= 1.7.1 this will return a tensor of ones, using the updated weights. Starting in PyTorch 1.8, this return a tensors of zeros.

The [documentation](https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html#torch.nn.Embedding) still states that:

"""
With padding_idx set, the embedding vector at padding_idx is initialized to all zeros. However, note that this vector can be modified afterwards, e.g., using a customized initialization method, and thus changing the vector used to pad the output. The gradient for this vector from Embedding is always zero.
"""

I have no idea if the change was intended and the documentation was not updated accordingly, or if the change was not intended in which case there is a bug to fix. If you could tell me which one it is, it would be much appreciated!

In the first case I think this would warrant a clear warning in the release notes as it was not obvious to me from reading them and I had a lot of pretrained models not working the same way in 🤗 Transformers when upgrading to 1.8.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @brianjo @mruberry

**Comment by zhangguanheng66** [comment]
I can reproduce this with torch 1.7.1 and 1.8.0. 

cc @albanD I don't see we have mentioned any relevant changes in the release note. Please correct me if I'm wrong.

**Comment by gchanan** [comment]
maybe https://github.com/pytorch/pytorch/pull/46714 or https://github.com/pytorch/pytorch/pull/47184?

**Comment by albanD** [comment]
Yep these two PRs are listed in the release notes as bug fix for the padding_idx behavior.

**Comment by zhangguanheng66** [comment]
Should we mark these as BC breakage in the release note and provide the instructions to adopt the change?

**Comment by sgugger** [comment]
I would argue that changing the documented behavior of a function (without adapting the documentation) does not characterize as bug fix but a breaking change and warrants a bit more warning to the users.

Just so I'm 100% clear on the plan forward and can adapt on the Hugging Face side, this change was intentional and will not be reverted?

**Comment by albanD** [comment]
@kshitij12345 is this the expected behavior?

**Comment by gchanan** [comment]
I confirmed the change happens in https://github.com/pytorch/pytorch/pull/46714.

**Comment by gchanan** [comment]
The new behavior looks wrong to me for the module.

The two snippets in the module documentation are:

> padding_idx (int, optional) – If given, pads the output with the embedding vector at padding_idx (initialized to zeros) whenever it encounters the index.


> With padding_idx set, the embedding vector at padding_idx is initialized to all zeros. However, note that this vector can be modified afterwards, e.g., using a customized initialization method, and thus changing the vector used to pad the output. The gradient for this vector from Embedding is always zero.

The new behavior clearly doesn't do this.

Now, I do think the new behavior fixes the _functional_ variant, which similarly claims:
> padding_idx (int, optional) – If given, pads the output with the embedding vector at padding_idx (initialized to zeros) whenever it encounters the index.

but previous examples just ignored the padding_idx.

**Comment by gchanan** [comment]
If we want versions that satisfy both sets of documentation and are _functional_ (don't modify the input), we need two different implementations.

**Comment by kshitij12345** [comment]
Looking at the documentation referred by @gchanan , it is clear that Module variant is broken.

Instead of two implementations, like @gchanan  suggested we can just pass `padding_idx=None` with a comment for the reason.

https://github.com/pytorch/pytorch/blob/369601355feccbbc35b428d5eba0313b4f7fc8fd/torch/nn/modules/sparse.py#L145-L147


This makes me wonder that if `from_pretrained` previously wanted to allow this behaviour and that's why it ignored padding_idx? (But it didn't match the documentation for `from_pretrained`).

**Comment by kshitij12345** [comment]
Looking at the backward implementation, if we pass `padding_idx=None` then that would be breaking,

```
The gradient for this vector from Embedding is always zero.
```

As gradient flows through all the touched vector except for the vector at `padding_idx`.
https://github.com/pytorch/pytorch/blob/369601355feccbbc35b428d5eba0313b4f7fc8fd/aten/src/ATen/native/Embedding.cpp#L107-L131

In light of this, second implementation which ignores the `padding_idx` (like before) makes more sense.

**Comment by jbschlosser** [comment]
Side note, but related. I find the documentation of `padding_idx` for the functional form to be strangely worded: "If given, pads the output with the embedding vector at padding_idx (initialized to zeros) whenever it encounters the index."

"initialized to zeros" doesn't really make sense in the functional context where `weight` is passed in. Imo the wording should be changed to indicate that the output is unconditionally padded with an all-zero embedding vector whenever `padding_idx` is encountered.

**Comment by jbschlosser** [comment]
I'm not convinced that either #46714 or #47184 are fully correct. Following is what I consider to be the "ideal semantics" for `padding_idx` that adhere to the principle of least astonishment, and the current state of the semantics in 1.8.0:

### Ideal semantics

 The crux of the issue is that `padding_idx` is used within `Embedding` to accomplish two things:
1. To "fix" the value of the embedding vector at `padding_idx` by preventing gradient updates to it
2. To pad the output with zeros wherever `padding_idx` is encountered in the input

The important part is 1, and 2 is unnecessary to handle separately because it can be achieved from 1. Ideally, `padding_idx` should only be used to indicate which embedding vector should be exempt from gradient updates. Otherwise, indexing should work as normal with no special handling needed - if the embedding vector at `padding_idx` is all zeros (AKA its initial value), normal indexing will produce the behavior of 2. It's important to note that 2 is not desired for all use cases!

This should be the case for both the module and functional forms, and it gives the user maximum flexibility to set the embedding vector at `padding_idx` to whatever is useful for their purposes without confusing semantics (arg naming aside).

### 1.8.0 semantics

`padding_idx` passed to the `Embedding` constructor:
* Embedding vector at `padding_idx` will be initialized to all zeros
* Embedding vector at `padding_idx` can be manually set to something else after initialization
* Embedding vector at `padding_idx` will not receive gradient updates
* Calling the module with `padding_idx` as an input gives the fixed embedding vector at `padding_idx`, whether it's all-zeros or a manually set value

`padding_idx` passed to `from_pretrained()`:
* Embedding vector at `padding_idx` will be overwritten with all zeros (changed due to #47184)
* Embedding vector at `padding_idx` can be manually set to something else **after loading**
* Embedding vector at `padding_idx` will not receive gradient updates
* Calling the module with `padding_idx` as an input gives either zeros or the manual value set **after loading** (changed due to #47184)
  * Note that the old behavior allowed for saving / loading a non-zero padding embedding vector. After #47184, if a non-zero padding embedding vector is saved, it will be overwritten with zeros when loaded. As a user, I'd find this surprising, although it is possible to work around by setting a manual value after loading

`padding_idx` passed to `F.embedding()`:
* Embedding vector at `padding_idx` will not receive gradient updates
* Calling the module with `padding_idx` as an input gives zeros unconditionally (changed due to #46714)
  * Note that the old behavior returned the embedding vector at `padding_idx` regardless of whether or not it was all zeros. After #46714, this changed to always produce zeros

### TL;DR

I think both #46714 and #47184 stray from the ideal `padding_idx` semantics and should not have been introduced. It seems they both came about as a way to reconcile with confusing documentation. Imo `padding_idx` should always have been documented as a way to set a specific embedding vector to a fixed value (default all zeros) that it is not affected by gradient updates.

**Comment by gchanan** [comment]
> mo padding_idx should always have been documented as a way to set a specific embedding vector to a fixed value (default all zeros) that it is not affected by gradient updates.

That seems to be the intention of the original PR: https://github.com/pytorch/pytorch/pull/6430 and issue: https://github.com/pytorch/pytorch/issues/6411.  Although I guess technically "the gradient is zero" is slightly different than "not affected by gradient updates", although not sure it matters.

**Comment by gchanan** [comment]
> Note that the old behavior allowed for saving / loading a non-zero padding embedding vector. After #47184, if a non-zero padding embedding vector is saved, it will be overwritten with zeros when loaded. As a user, I'd find this surprising, although it is possible to work around by setting a manual value after loading

I agree this seems surprising.

**Comment by kshitij12345** [comment]
https://github.com/pytorch/pytorch/issues/53368#issuecomment-793284794

Agree with @jbschlosser .
Will revert the changes and update the documentation to better capture the behaviour of `padding_idx` in #53447
